src/crimson/osd/shard_services.cc

```diff
@@ -211,7 +211,16 @@ seastar::future<> ShardServices::send_pg_temp()
     }
     m->pg_temp.emplace(pgid, pg_temp.acting);
   }
-  return seastar::parallel_for_each(std::begin(ms), std::end(ms),
+  std::vector<seastar::future<>> fut_vec;
+  for (auto& m : ms) {
+    if (m) {
+      fut_vec.emplace_back(monc.send_message(m));
+    }
+  }
+  _sent_pg_temp();
+  return when_all_succeed(std::begin(fut_vec), std::end(fut_vec))
```

Comment:
i don't quite understand this change. `ShardServices::_sent_pg_temp()` is called after **all** `MOSDPGTemp` are sent in existing implementation. what's wrong with it?

Reply:
There might be such cases like:
1. parallel_for_each sending all pgtemps to monitors
2. new pgtemps created
3. _sent_pg_temp is called
In the above case, pgtemps created in step 2 would be erased be _sent_pg_temp and wouldn't get the chance to be sent to monitors

Reply:
how about moving `_sent_pg_temp()` up after the `for` loop.

Reply:
Do you mean this?

```
auto fut = parallel_for_each(std::begin(ms), std::end(ms), [] {...});
_sent_pg_temp();
return fut;

```

If so, I agree with you:-)